source/common/upstream/load_balancer_impl.h

```diff
@@ -216,64 +216,118 @@ class ZoneAwareLoadBalancerBase : public LoadBalancerBase {
+    ASSERT(rr_indexes_.find(source) != rr_indexes_.end());
+    return hosts_to_use[rr_indexes_[source]++ % hosts_to_use.size()];
+  }
+
+  std::unordered_map<HostsSource, uint64_t, HostsSourceHash> rr_indexes_;
 };
 
 /**
  * Weighted Least Request load balancer.
  *
  * In a normal setup when all hosts have the same weight of 1 it randomly picks up two healthy hosts
- * and compares number of active requests.
- * Technique is based on http://www.eecs.harvard.edu/~michaelm/postscripts/mythesis.pdf
+ * and compares number of active requests. Technique is based on
+ * http://www.eecs.harvard.edu/~michaelm/postscripts/mythesis.pdf and is known as P2C (power of
+ * two choices).
  *
- * When any of the hosts have non 1 weight, apply random weighted balancing.
- * Randomly pickup the host and send 'weight' number of requests to it.
- * This technique is acceptable for load testing but
- * will not work well in situations where requests take a long time.
- * In that case a different algorithm using a full scan will be required.
+ * When any hosts have a weight that is not 1, an RR EDF schedule is used. Host weight is scaled
+ * by the number of active requests at pick/insert time. Thus, hosts will never fully drain as
+ * they would in normal P2C, though they will get picked less and less often. In the future, we
+ * can consider two alternate algorithms:
+ * 1) Expand out all hosts by weight (using more memory) and do standard P2C.
+ * 2) Use a weighted Maglev table, and perform P2C on two random hosts selected from the table.
+ *    The benefit of the Maglev table is at the expense of resolution, memory usage is capped.
+ *    Additionally, the Maglev table can be shared amongst all threads.
  */
-class LeastRequestLoadBalancer : public LoadBalancer, ZoneAwareLoadBalancerBase {
+class LeastRequestLoadBalancer : public EdfLoadBalancerBase {
 public:
   LeastRequestLoadBalancer(const PrioritySet& priority_set, const PrioritySet* local_priority_set,
                            ClusterStats& stats, Runtime::Loader& runtime,
                            Runtime::RandomGenerator& random,
-                           const envoy::api::v2::Cluster::CommonLbConfig& common_config);
-
-  // Upstream::LoadBalancer
-  HostConstSharedPtr chooseHost(LoadBalancerContext* context) override;
+                           const envoy::api::v2::Cluster::CommonLbConfig& common_config)
+      : EdfLoadBalancerBase(priority_set, local_priority_set, stats, runtime, random,
+                            common_config) {
+    initialize();
+  }
 
 private:
-  HostSharedPtr last_host_;
-  uint32_t hits_left_{};
+  void refreshHostSource(const HostsSource&) override {}
+  double hostWeight(const Host& host) override {
+    // Here we scale host weight by the number of active requests at the time we do the pick. We
+    // always add 1 to avoid division by 0.
+    return static_cast<double>(host.weight()) / (host.stats().rq_active_.value() + 1);
```

Comment:
One interesting sanity check is to ask "if we used the weighted algorithm in the case of all weights being equal to 1, would we get the same behavior as power-of-two balancer, at least statistically?" Does this have this property?

Reply:
I think not, mostly due to the draining semantics that I already commented about. Though even beyond that I'm not sure the math works out the same. Did you want more comments here along those lines?